src/pem_decoder.rs

```diff
@@ -0,0 +1,200 @@
+      },
+      Err(_) => return Err(ErrorKind::InvalidKeyFormat)?,
+    }
+  }
+
+  /// This will do the initial parsing of a PEM file.
+  /// Supported tagged pems include "RSA PRIVATE KEY", "RSA PUBLIC KEY",
+  /// "PRIVATE KEY", "PUBLIC KEY"
+  /// PEMs with multiple tagged portions are not supported
+  pub fn as_key(&self) -> Result<Key<'_>> {
+    match self.encoded_with {
+      PemEncodedWith::PKCS1 => Ok(Key::Der(self.content.as_slice())),
+      PemEncodedWith::PKCS8 => {
+        match self.pem_type {
+          PemType::RSAPrivateKey => Ok(Key::Der(extract_first_bitstring(&self.asn1)?)),
+          PemType::RSAPublicKey => Ok(Key::Der(extract_first_bitstring(&self.asn1)?)),
+          PemType::ECPrivateKey => Ok(Key::Pkcs8(self.content.as_slice())),
+          PemType::ECPublicKey => Ok(Key::Pkcs8(extract_first_bitstring(&self.asn1)?)),
+        }
+      },
+    }
+  }
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+/// Supported PEM files for EC and RSA Public and Private Keys
+enum PemType {
+  ECPublicKey,
+  ECPrivateKey,
+  RSAPublicKey,
+  RSAPrivateKey,
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+enum PemEncodedWith {
+  PKCS1,
+  PKCS8,
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+enum Classification {
+  EC,
+  RSA,
+}
+
+// This really just finds and returns the first bitstring or octet string
+// Which is the x coordinate for EC public keys
+// And the DER contents of an RSA key
+// Though PKCS#11 keys shouldn't have anything else. 
+// It will get confusing with certificates.
+fn extract_first_bitstring(asn1: &Vec<simple_asn1::ASN1Block>) -> Result<&[u8]> {
+  for asn1_entry in asn1.iter() {
+    match asn1_entry {
+      simple_asn1::ASN1Block::Sequence(_, entries) => {
+        if let Ok(result) = extract_first_bitstring(entries) {
+          return Ok(result);
+        }
+      }
+      simple_asn1::ASN1Block::BitString(_, _, value) => {
+        return Ok(value.as_ref());
+      }
+      simple_asn1::ASN1Block::OctetString(_, value) => {
+        return Ok(value.as_ref());
+      }
+      _ => ()
+    }
+  }
+  return Err(ErrorKind::InvalidEcdsaKey)?
+}
+
+fn classify_pem(asn1: &Vec<simple_asn1::ASN1Block>) -> Option<Classification> {
+  // These should be constant but the macro requires
+  // #![feature(const_vec_new)]
+  let ec_public_key_oid = simple_asn1::oid!(1,2,840,10045,2,1);
+  let rsa_public_key_oid = simple_asn1::oid!(1,2,840,113549,1,1,1);
+
+  for asn1_entry in asn1.iter() {
+    match asn1_entry {
+      simple_asn1::ASN1Block::Sequence(_, entries) => {
+        if let Some(classification) = classify_pem(entries) {
+          return Some(classification);
+        }
+      }
+      simple_asn1::ASN1Block::ObjectIdentifier(_, oid) => {
+        if oid == ec_public_key_oid {
+          return Option::Some(Classification::EC);
+        } else if oid == rsa_public_key_oid {
+          return Option::Some(Classification::RSA);
+        }
+      }
+      _ => {
+        println!("Ignoring {:?}", asn1_entry);
```

Comment:
you can remove that

Reply:
oh yes definitely, that was a debugging helper.